feat: Display vCard contact name in the message summary

iequidoo opened the pull request.

iequidoo commented:
We even can remove `StockMessage::Contact` now, it should never be displayed

Hocuri reviewed (commented):
I'm a bit concerned that this makes `Viewtype::Vcard` very special code-wise because `set_file()` changes the viewtype to `Viewtype::File` or sets a param for only this one viewtype, which could produce subtle bugs in the future.

Would it be feasible put the code calling `try_set_vcard()` into `do_set_draft()` and `prepare_msg_blob()` instead of `set_file()` and `set_file_from_bytes()` (and keep the code in `mimeparser.rs`)?

Note to myself, I need to test this on an Android device before approving - for [vcards created with DC, vcards created with Thunderbird, vcards created with Thunderbird that contain too many and should be shown as a file]:
- send .vcf file (sender & receiver side)
- draft file
- forward file
- share file
- receive file

iequidoo commented:
> I'm a bit concerned that this makes `Viewtype::Vcard` very special code-wise because `set_file()` changes the viewtype to `Viewtype::File` or sets a param for only this one viewtype, which could produce subtle bugs in the future.
> 
> Would it be feasible put the code calling `try_set_vcard()` into `do_set_draft()` and `prepare_msg_blob()` instead of `set_file()` and `set_file_from_bytes()` (and keep the code in `mimeparser.rs`)?

It also was a question for me where it's better to parse a vCard on the sender side (and fall back to `File` if it can't be parsed). Pros of doing this in `set_file*()` are:
- `set_file_from_bytes()` already has data in memory, no need to open the file later.
- `test_get_summary_text()` checks that the summary is correct already after `set_file_from_bytes()`, no need to make vCards a special case.

Otoh it can be unexpected that `set_file*()` may change the message viewtype. Btw, also if the file content changes after the `set_file*()` call, the summary will be wrong.

Wrt the code in `mimeparser.rs`, in fact i just moved the check for vCard validity from `get_mime_type()` to `do_add_single_file_part()`, this way we don't need to parse a vCard twice -- the first time to check its validity and the second to set the message summary.

iequidoo commented on src/mimeparser.rs:
```diff
@@ -1937,7 +1946,7 @@ fn get_mime_type(
 
     let viewtype = match mimetype.type_() {
         mime::TEXT => match mimetype.subtype() {
-            mime::VCARD if is_valid_deltachat_vcard(mail) => Viewtype::Vcard,
+            mime::VCARD => Viewtype::Vcard,
```
On a receiver side the viewtype is fixed in `do_add_single_file_part()`. Btw, it already has some magic for `Viewtype::Webxdc`, so i think it's ok to check vCards there

Hocuri replied:
I see. So, while this still seems unnecessarily confusing to me, the PR here is in line with the current code.

Could you change the doc comment of `get_mime_type()` to mention this? E.g.:

```rust
/// Return the mimetype and viewtype for a parsed mail.
/// 
/// This only looks at the metadata, not at the content;
/// for some viewtypes like `Viewtype::Vcard` and `Viewtype::Webxdc`,
/// the viewtype may later be corrected in `do_add_single_file_part()`.
```

iequidoo replied:
Added your comment, but w/o listing specific viewtypes. For `Webxdc` btw it's vice versa -- `File` may be changed to `Webxdc` later.

iequidoo commented:
> We even can remove `StockMessage::Contact` now, it should never be displayed

Decided not to deprecate it yet, it may be useful in the future, but added a comment that it's unused currently

Hocuri commented:
> `set_file_from_bytes()` already has data in memory, no need to open the file later.

Except for Webxdcs, `set_file_from_bytes()` is only used in tests where performance doesn't matter as much, though.

> `test_get_summary_text()` checks that the summary is correct already after `set_file_from_bytes()`, no need to make vCards a special case.

I think it's better to special-case in a test than to special-case `set_file*()` to do something extra just for a single viewtype.

> > We even can remove `StockMessage::Contact` now, it should never be displayed
> 
> Decided not to deprecate it yet, it may be useful in the future, but added a comment that it's unused currently

I do think we should remove it, if we add stock strings just in case we may need it in the future, we will end up with a lot of unused stock strings at some point. And in most cases when one thinks "it may be useful in the future", one ends up never needing it.

iequidoo commented:
> I think it's better to special-case in a test than to special-case `set_file*()` to do something extra just for a single viewtype.

Moved updating the message summary to `prepare_msg_blob()` and `do_set_draft()`.

> I do think we should remove it, if we add stock strings just in case we may need it in the future, we will end up with a lot of unused stock strings at some point. And in most cases when one thinks "it may be useful in the future", one ends up never needing it.

Removed.

iequidoo commented:
`DC_STR_CONTACT` can't be just removed so i deprecated it.

Hocuri commented on src/mimeparser.rs:
```diff
@@ -1988,15 +2000,17 @@ fn is_attachment_disposition(mail: &mailparse::ParsedMail<'_>) -> bool {
+pub(crate) fn get_vcard_summary(vcard: &[u8]) -> Option<String> {
+    let vcard = str::from_utf8(vcard).ok()?;
+    let contacts = deltachat_contact_tools::parse_vcard(vcard);
+    let [c] = &contacts[..] else {
+        return None;
     };
-    let contacts = deltachat_contact_tools::parse_vcard(body);
-    if let [c] = &contacts[..] {
-        return deltachat_contact_tools::may_be_valid_addr(&c.addr);
+    if !deltachat_contact_tools::may_be_valid_addr(&c.addr) {
+        return None;
     }
-    false
+    Some(c.display_name().to_string())
```
Nit: This function could be moved to `message.rs`, I think it would fit there slightly better, but feel free to ignore

Hocuri commented on src/summary.rs:
```diff
@@ -385,7 +394,8 @@ mod tests {
         .await
         .unwrap();
-        assert_summary_texts(&msg, ctx, "👤 Contact").await;
+        chat_id.set_draft(ctx, Some(&mut msg)).await.unwrap();
+        assert_summary_texts(&msg, ctx, "👤 Alice Wonderland").await;
```
Would be nice to also test somewhere that the summary is shown correctly after sending it (`ctx.send_msg(chat_id, &mut msg)`)

iequidoo replied:
Added a check to `test_receive_vcard()` as it already sends vCards

Hocuri reviewed (commented):
Nice! I have 3 small, non-blocking comments. The only thing left to approve is me testing that this works, as mentioned above.

Hocuri commented:
Status of this PR: When testing on a real Android device, I found that this doesn't work on the sending side and the summary is only "👤". @iequidoo and me texted in private about this, but we don't know why it happens (and I still have somewhat limited capacity to debug it).

iequidoo commented:
> Status of this PR: When testing on a real Android device, I found that this doesn't work on the sending side and the summary is only "👤". @iequidoo and me texted in private about this, but we don't know why it happens (and I still have somewhat limited capacity to debug it).

I tried to debug it on my Android yesterday, but had a problem with the file system and unfortunately lost my build env :) But i have reproduced the problem before that. Now building it again.

iequidoo commented:
> I tried to debug it on my Android yesterday, but had a problem with the file system and unfortunately lost my build env :) But i have reproduced the problem before that. Now building it again.

Now it works on my Android, so the last commit is really a fix.

iequidoo commented:
I've fixed it for drafts, added the corresponding test and verified on my Android

iequidoo commented on src/chat.rs:
```diff
@@ -894,8 +893,15 @@ impl ChatId {
                 msg.param.set(Param::File, blob.as_name());
-                if blob.suffix() == Some(WEBXDC_SUFFIX) {
-                    msg.viewtype = Viewtype::Webxdc;
+                if msg.viewtype == Viewtype::File {
+                    if let Some((better_type, _)) =
+                        message::guess_msgtype_from_suffix(&blob.to_abs_path())
+                    {
+                        msg.viewtype = better_type;
+                    }
+                }
```
True, and we don't have a test for this scenario, i'm going to add it then. Also not to break things much i suggest to correct the viewtype here only to `Webxdc` and `Vcard`, remaining corrections are done by `prepare_msg_blob()` anyway

iequidoo replied:
Fixed and created a separate PR for the corresponding test as it looks somewhat unrelated to this: https://github.com/deltachat/deltachat-core-rust/pull/5685

Hocuri commented on src/chat.rs:
```diff
@@ -894,8 +893,16 @@ impl ChatId {
+                if msg.viewtype == Viewtype::File {
+                    if let Some((better_type, _)) =
+                        message::guess_msgtype_from_suffix(&blob.to_abs_path())
+                            .filter(|&(vt, _)| vt == Viewtype::Webxdc || vt == Viewtype::Vcard)
```
Suggestion, non-blocking: Could this get a comment like "If the attached file is a VCard or a Webxdc, automatically convert it to a VCard/Webxdc. We do not do this for other files, because we recommend that users can send images as a “file” to preserve the original quality (usually, we compress some files like images)."?

iequidoo replied:
Added the comment, only slightly changed it

Hocuri commented on src/mimeparser.rs:
```diff
@@ -1233,6 +1234,7 @@ impl MimeMessage {
                 return Ok(());
             }
         }
+        let mut param = Params::new();
```
Nit, feel free to ignore: With this chnange, if someone should write `part.param.set(...)` instead of `param.set(...)`, then the param wouldn't be changed at all because it will later be overwritten by the line `part.param = param;`

I would very slightly prefer to write `let mut part = Part::default();` up here already, and to write `part.param.set(...)` throughout the function, in order to avoid having this duplicate `Param` (one in the `param` variable and one inside the `part`).
